MetaMask/metamask-extension
# NetworkController: Split `network` into `networkId` and `networkStatus`

## Explanation



The `network` store of the network controller crams two types of data into one place. It roughly tracks whether we have enough information to make requests to the network and whether the network is capable of receiving requests, but it also stores the ID of the network (as obtained via `net_version`).

Generally we shouldn't be using the network ID for anything, as it has been completely replaced by chain ID, which all custom RPC endpoints have been required to support for over a year now. However, as the network ID is used in various places within the extension codebase, removing it entirely would be a non-trivial effort. So, minimally, this commit splits `network` into two stores: `networkId` and `networkStatus`. But it also expands the concept of network status.

Previously, the network was in one of two states: "loading" and "not-loading". But now it can be in one of four states:

- `available`: The network is able to receive and respond to requests.
- `unavailable`: The network is not able to receive and respond to requests for unknown reasons.
- `blocked`: The network is actively blocking requests based on the user's geolocation. (This is specific to Infura.)
- `unknown`: We don't know whether the network can receive and respond to requests, either because we haven't checked or we tried to check and were unsuccessful.

This commit also changes how the network status is determined — specifically, how many requests are used to determine that status, when they occur, and whether they are awaited. Previously, the network controller would make 2 to 3 requests during the course of running `lookupNetwork`.

* First, if it was an Infura network, it would make a request for `eth_blockNumber` to determine whether Infura was blocking requests or not, then emit an appropriate event. This operation was not awaited.
* Then, regardless of the network, it would fetch the network ID via `net_version`. This operation was awaited.
* Finally, regardless of the network, it would fetch the latest block via `eth_getBlockByNumber`, then use the result to determine whether the network supported EIP-1559. This operation was awaited.

Now:

* One fewer request is made, specifically `eth_blockNumber`, as we don't need to make an extra request to determine whether Infura is blocking requests; we can reuse `eth_getBlockByNumber`;
* All requests are awaited, which makes `lookupNetwork` run fully in-band instead of partially out-of-band; and
* Both requests for `net_version` and `eth_getBlockByNumber` are performed in parallel to make `lookupNetwork` run slightly faster.

Fixes #16946.

## Manual Testing Steps



This PR touches a lot of files, but here are some pointers:

* If you run `yarn start` on this branch and restart your local version of MetaMask, if you were previously connected to a network, you should still be connected to that network when you restart the extension
* The loading spinner should spin when you load MetaMask for the first time and the network is being connected to
* Switching networks should continue to work as intended (should be able to send a transaction and otherwise interact with a network, switch network, and interact with the other network in the same way)
* Pending transactions should complete in the order that they were added to the queue
* Should be able to perform a swap on one network, switch the network, and perform a swap on the other network
* There should be no errors printed to the console when first loading, and then interacting, with the test dapp

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
